### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by dionys93** [dionys93/ety-dictionary#86]
<html><head></head><body><h1>Roadmap: teaching an AI the Inglisce orthography</h1>
<p><em>Converted from an open issue. Roadmaps filed as issues never close — they accumulate
comments and drift out of date while staying open, and they can't be linked to from the
individual issues they spawn. This belongs at <code>toolkit/docs/ROADMAP.md</code>, versioned
alongside <code>ADR.md</code>, with each phase filed as its own issue when it becomes actionable.</em></p>
<hr>
<h2>The core advantage</h2>
<p>The architecture provides something most conlang projects lack: a <strong>deterministic oracle</strong>.
The pipeline can generate unlimited English↔Inglisce parallel data and grade a model's
output automatically. Everything below is built around that.</p>
<p>A framing note that shapes the whole plan: because transcription is word-level
relexification plus morphology with English syntax preserved, the learning task is closer
to "cipher plus suffix rules" than to language acquisition. That is very learnable, and it
means in-context approaches may get further than expected before fine-tuning is needed.</p>
<hr>
<h2>Phase 0 — Data integrity <em>(new; blocks everything)</em></h2>
<p>Not in the original plan, added because the corpus audit changed the picture.</p>
<p>Phase 1 originally opened with "fix items 1, 3 and 5 so the data is trustworthy." That
instinct was right and the scope was too small. Current state of <code>master_dataset.jsonl</code>
(8,839 stanzas, 7,108 unique headwords):</p>

Problem | Count
-- | --
Reformed spellings stored decomposed (NFD) vs composed (NFC) | 3,080 / 170
Headwords carrying more than one reformed spelling | 632
Etymology lines using [VL] and other excluded reconstructed languages | 168
Language tags in the data absent from config/languages.tsv | 56
Stanzas with an empty language tag | 55
Stanzas with fewer than two sources | 2,970 (34%)

<p>Two of these are fatal to the later phases specifically:</p>
<p><strong>The normalisation split.</strong> Phase 5 already anticipates that byte-level tokenizers
fragment <code>þ</code> and <code>Ћ</code> into multiple tokens. Mixed NFC/NFD makes that materially worse: the
same visible word tokenises two different ways depending on which stanza it came from, so
the model sees <code>tône</code> and <code>tône</code> as unrelated strings and has to learn both. Every phase
downstream inherits this, and the fix gets more expensive at each one.</p>
<p><strong>The 632 spelling conflicts.</strong> These are the orthography contradicting itself — <code>tâme</code>/<code>tâm</code>,
<code>þainche</code>/<code>þainc</code>, <code>traice</code>/<code>traisse</code>. A model trained on both learns that the rules are
probabilistic when they are meant to be deterministic, and the eval harness in Phase 4 has
no stable ground truth to grade against.</p>
<ul>
<li>[ ] Settle canonical Unicode normalisation and run a one-time pass over the corpus</li>
<li>[ ] Resolve the 632 spelling conflicts</li>
<li>[ ] Reconcile <code>config/languages.tsv</code> with the working tag list</li>
<li>[ ] Add <code>etym-lint</code> rules for all of the above so none of it recurs</li>
</ul>
<p>See the audit notebook issue for how these were found and how the worklists are generated.</p>
<hr>
<h2>Phase 1 — Harden and scale the corpus</h2>
<p>Drive the missing-words loop: translate books → export missing words → fill dictionary →
retranslate. Coverage is the bottleneck for everything downstream.</p>
<ul>
<li>[ ] Add a stats step reporting translated-token percentage per book, so coverage can be
watched climbing</li>
<li>[ ] Fix the contraction bug first (see below) — it silently corrupts output rather than
reporting a miss, so it won't show up in the missing-words tracker</li>
</ul>
<p><strong>Note on the trackers.</strong> The README documents <code>toolkit/dist/missing_words.txt</code> and
<code>fallback_words.txt</code>, but neither is present in <code>dist/</code>. Either the pipeline hasn't been
run since they were added, or they're being written elsewhere. Worth confirming before
relying on the loop.</p>
<p><strong>Note on contractions.</strong> <code>transcriber.js</code> bypasses clitic tokens as literal English, but
spaCy splits <code>can't</code> into <code>ca</code> + <code>n't</code>. The clitic passes through untranslated and the stem
resolves via lemma <code>can</code> to the affirmative modal, so negation is dropped without a miss
being recorded. The 16 contraction keys the parser correctly compiles into
<code>translationBrain.json</code> are unreachable. This inflates apparent coverage and poisons any
parallel corpus built before it's fixed.</p>
<hr>
<h2>Phase 2 — Export a sentence-aligned parallel corpus</h2>
<p>Because transcription is token-by-token, alignment is free. What's needed is sentence
segmentation.</p>
<ul>
<li>[ ] Add sentence indices to the AST. <code>spacy_parser.py</code> currently maps over <code>doc</code> tokens
only — <code>extract_token_data</code> returns <code>text</code>, <code>lemma</code>, <code>pos</code>, <code>tag</code>, <code>is_ent</code>,
<code>whitespace</code>, with no sentence membership. <code>doc.sents</code> is available; this is a
one-field addition to the extractor.</li>
<li>[ ] Emit <code>{"en": "...", "ing": "..."}</code> JSONL pairs</li>
<li>[ ] Filter out pairs containing <code>[bracketed]</code> missing words</li>
<li>[ ] Report retained-pair count per book</li>
</ul>
<p>This corpus is the asset every later phase depends on.</p>
<hr>
<h2>Phase 3 — Language pack for in-context learning</h2>
<p>An exporter that compiles three things:</p>
<ul>
<li>[ ] <strong>Orthography and morphology spec in prose</strong> — <code>utils.js</code> and the parser's six
conjugation classes rendered as rules a model can read, including <code>þ</code>/<code>Ћ</code>, the
accented vowels, and casing conventions</li>
<li>[ ] <strong>Frequency-ranked core lexicon</strong> — the full brain is 7,118 lemma keys and won't fit
a context window; top 1–3k by book frequency will</li>
<li>[ ] <strong>50–100 diverse parallel sentence pairs</strong> from Phase 2</li>
</ul>
<p>Load into a Claude Project or system prompt. Modern models apply a documented
relexification in-context reasonably well, and this costs an export script and nothing else.</p>
<p>Worth including in the spec: the affix reflex table, once #3 settles it. English <code>-ion</code>
maps to <code>-cion</code>, <code>-sion</code>, <code>-tion</code> and <code>-ione</code> in the reformed orthography, and a model that
doesn't know the conditioning will guess.</p>
<hr>
<h2>Phase 4 — Eval harness, before optimising anything</h2>
<ul>
<li>[ ] Hold out a test set of sentence pairs, excluded from the language pack</li>
<li>[ ] Model translates English → Inglisce</li>
<li>[ ] Score against pipeline output: exact match plus chrF</li>
</ul>
<p>Character-level metrics matter here because the orthography lives in the suffixes, and a
model that gets every lemma right and every suffix wrong scores well on word-level metrics.
Fully automatable, since the pipeline is ground truth.</p>
<p>This phase answers whether Phase 5 is needed at all. Build it before assuming it is.</p>
<hr>
<h2>Phase 5 — Fine-tune, only if evals say so</h2>
<p>If in-context performance plateaus:</p>
<ul>
<li>[ ] Format the parallel corpus as instruction data, both directions</li>
<li>[ ] Generate "explain why this suffix" examples from conjugation metadata</li>
<li>[ ] LoRA-tune an open-weights model (Llama, Qwen). Low thousands of pairs minimum, which
Phase 1 should comfortably provide</li>
<li>[ ] Mix in plain-English data to avoid degrading the base model</li>
</ul>
<p>Tokenizer note, revised from the original: byte-level tokenizers handle <code>þ</code> and <code>Ћ</code> fine
but fragment them. <strong>This is why Phase 0 is a hard blocker</strong> — with mixed normalisation the
fragmentation is not merely inefficient but inconsistent, and the model has to learn the
same word twice.</p>
<hr>
<h2>Phase 6 — Close the loop</h2>
<ul>
<li>[ ] Round-trip validation: model output back through the dictionary to English</li>
<li>[ ] Feed systematic errors into the language pack as targeted counter-examples</li>
<li>[ ] Track chrF over successive language-pack revisions</li>
</ul>
<hr>
<h2>Dependency summary</h2>
<pre><code>Phase 0  (data integrity)
   └─&gt; Phase 1  (coverage; needs the contraction fix)
          └─&gt; Phase 2  (parallel corpus; needs doc.sents in the AST)
                 └─&gt; Phase 3  (language pack; wants the affix reflex table)
                        └─&gt; Phase 4  (evals)
                               └─&gt; Phase 5  (fine-tune, conditional on Phase 4)
                                      └─&gt; Phase 6  (round-trip)
</code></pre>
<p>Phase 0 is the only one that can start today, and skipping it makes every later phase more
expensive rather than merely later.</p></body></html>
